File: ios/PouleParty/Components/RoadSnapService.swift
//
//  RoadSnapService.swift
//  PouleParty
//
//  Snaps coordinates to the nearest walkable road using the Mapbox
//  Directions API. Each waypoint is independently snapped to the
//  nearest road segment. Falls back to the original position when
//  the API is unreachable or no snap is found.
//

import CoreLocation
import Foundation
import os

private let logger = Logger(category: "RoadSnapService")

enum RoadSnapService {

    /// Snaps an array of coordinates to the nearest walkable roads.
    /// Processes each point individually via the Directions API to
    /// guarantee independent snapping (no route dependency).
    static func snapToRoads(_ coordinates: [CLLocationCoordinate2D]) async -> [CLLocationCoordinate2D] {
        guard !coordinates.isEmpty,
              let token = mapboxAccessToken else {
            return coordinates
        }

        return await withTaskGroup(of: (Int, CLLocationCoordinate2D).self) { group in
            for (index, coord) in coordinates.enumerated() {
                group.addTask {
                    let snapped = await snapSinglePoint(coord, accessToken: token)
                    return (index, snapped)
                }
            }

            var result = coordinates
            for await (index, snapped) in group {
                result[index] = snapped
            }
            return result
        }
    }

    // MARK: - Private

    /// Snaps a single coordinate to the nearest walkable road.
    /// Uses the Directions API with a tiny offset point so the API
    /// returns the snapped waypoint location.
    private static func snapSinglePoint(
        _ coordinate: CLLocationCoordinate2D,
        accessToken: String
    ) async -> CLLocationCoordinate2D {
        // Create a second point ~10m north so the Directions API has a valid route
        let offsetLat = coordinate.latitude + 0.0001 // ~11m north
        let coordString = "\(coordinate.longitude),\(coordinate.latitude);\(coordinate.longitude),\(offsetLat)"

        let urlString = "https://api.mapbox.com/directions/v5/mapbox/walking/\(coordString)"
            + "?access_token=\(accessToken)"
            + "&overview=false"
            + "&steps=false"

        guard let url = URL(string: urlString) else { return coordinate }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                logger.warning("Directions API returned non-200 status")
                return coordinate
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let waypoints = json["waypoints"] as? [[String: Any]],
                  let firstWaypoint = waypoints.first,
                  let location = firstWaypoint["location"] as? [Double],
                  location.count == 2 else {
                return coordinate
            }

            let snapped = CLLocationCoordinate2D(latitude: location[1], longitude: location[0])

            // Reject snaps that moved the point too far (> 200m) — likely a data issue
            let original = CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
            let snappedLoc = CLLocation(latitude: snapped.latitude, longitude: snapped.longitude)
            if original.distance(from: snappedLoc) > 200 {
                logger.info("Snap moved point > 200m, keeping original")
                return coordinate
            }

            return snapped

        } catch {
            logger.warning("Road snap failed: \(error.localizedDescription)")
            return coordinate
        }
    }

    private static var mapboxAccessToken: String? {
        Bundle.main.object(forInfoDictionaryKey: "MBXAccessToken") as? String
    }
}

File: android/app/src/main/java/dev/rahier/pouleparty/ui/gamelogic/RoadSnapService.kt
package dev.rahier.pouleparty.ui.gamelogic

import android.util.Log
import com.mapbox.geojson.Point
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Snaps coordinates to the nearest walkable road using the Mapbox
 * Directions API. Each waypoint is independently snapped to the
 * nearest road segment. Falls back to the original position when
 * the API is unreachable or no snap is found.
 */
object RoadSnapService {

    private const val TAG = "RoadSnapService"
    private const val MAX_SNAP_DISTANCE_METERS = 200.0

    /**
     * Snaps a list of points to the nearest walkable roads.
     * Processes each point individually to guarantee independent snapping.
     */
    suspend fun snapToRoads(points: List<Point>, accessToken: String): List<Point> {
        if (points.isEmpty()) return points

        return coroutineScope {
            points.mapIndexed { _, point ->
                async { snapSinglePoint(point, accessToken) }
            }.awaitAll()
        }
    }

    /**
     * Snaps a single point to the nearest walkable road.
     * Uses the Directions API with a tiny offset point so the API
     * returns the snapped waypoint location.
     */
    private suspend fun snapSinglePoint(point: Point, accessToken: String): Point {
        return withContext(Dispatchers.IO) {
            try {
                // Create a second point ~10m north so the Directions API has a valid route
                val offsetLat = point.latitude() + 0.0001 // ~11m north
                val coordString = "${point.longitude()},${point.latitude()};${point.longitude()},$offsetLat"

                val urlString = "https://api.mapbox.com/directions/v5/mapbox/walking/$coordString" +
                    "?access_token=$accessToken" +
                    "&overview=false" +
                    "&steps=false"

                val connection = URL(urlString).openConnection() as HttpURLConnection
                try {
                    connection.requestMethod = "GET"
                    connection.connectTimeout = 5_000
                    connection.readTimeout = 5_000

                    if (connection.responseCode != 200) {
                        Log.w(TAG, "Directions API returned ${connection.responseCode}")
                        return@withContext point
                    }

                    val body = connection.inputStream.bufferedReader().use { it.readText() }
                    val json = JSONObject(body)
                    val waypoints = json.getJSONArray("waypoints")

                    if (waypoints.length() == 0) return@withContext point

                    val firstWaypoint = waypoints.getJSONObject(0)
                    val location = firstWaypoint.getJSONArray("location")
                    val snapped = Point.fromLngLat(location.getDouble(0), location.getDouble(1))

                    // Reject snaps that moved the point too far (> 200m)
                    val distance = haversineDistance(
                        point.latitude(), point.longitude(),
                        snapped.latitude(), snapped.longitude()
                    )
                    if (distance > MAX_SNAP_DISTANCE_METERS) {
                        Log.i(TAG, "Snap moved point > 200m, keeping original")
                        return@withContext point
                    }

                    snapped
                } finally {
                    connection.disconnect()
                }
            } catch (e: Exception) {
                Log.w(TAG, "Road snap failed, using original position", e)
                point
            }
        }
    }

    private fun haversineDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val r = 6_371_000.0 // Earth radius in meters
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
            Math.sin(dLon / 2) * Math.sin(dLon / 2)
        return r * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
    }
}
